service/javadsl/api/src/main/scala/com/lightbend/lagom/internal/javadsl/api/ServiceReader.scala

```diff
@@ -55,9 +55,14 @@ object ServiceReader {
           // Now using unreflect special, we get the default method from the declaring class, rather than the proxy
           .unreflectSpecial(method, method.getDeclaringClass)
     } else {
-      val lookup     = MethodHandles.lookup
-      val returnType = MethodType.methodType(classOf[Descriptor])
-      method => lookup.findSpecial(method.getDeclaringClass, method.getName, returnType, method.getDeclaringClass)
+      val lookup = MethodHandles.lookup
+      method =>
+        lookup.findSpecial(
+          method.getDeclaringClass,
+          method.getName,
+          MethodType.methodType(method.getReturnType, method.getParameterTypes),
+          method.getDeclaringClass
+        )
```

Comment:
I may be in need of more coffee, but please clarify the consequences of this. 

I'm understand it as following: 

The reader scans the type for default impl. Before, you were only expected to have one default method returning `Descriptor`. 

Now, it will be able to generated handles for all kinds of default method it finds. However, the `ServiceReader` should only be looking for the once returning `Descriptor` and with zero param. Correct?

Reply:
@gnuzzz please, share more information about your research with calling of proxy and caught exception.

Reply:
While reading service descriptor, the reader creates service proxy stub and calls `descriptor` method on it (see `ServiceReader#readServiceDescriptor`):
```
...
  val serviceStub =
        JProxy.newProxyInstance(classLoader, Array(serviceInterface), invocationHandler).asInstanceOf[Service]
  serviceStub.descriptor()
...
```
If there is any non-static calls from `descriptor`, it is a proxy stub call and passes through `ServiceReader.ServiceInvocationHandler#invoke`.
And then if called method is default, invocation handler tries to create method handle:
```
  ...
  if (method.isDefault) {
    // This is the way to invoke default methods via reflection, using the JDK7 method handles API
    defaultMethodHandleSupplier(method)
...
```
In current implementation of `defaultMethodHandleSupplier` there is different behaviour for java 8 and other java versions - for java 8 it can create method hadle for any default method (regardless of return type and parameter types):
```
  if (isJava8) {
      val methodHandlesLookupConstructor =
        classOf[MethodHandles.Lookup].getDeclaredConstructor(classOf[Class[_]], classOf[Int])
      if (!methodHandlesLookupConstructor.isAccessible)
        methodHandlesLookupConstructor.setAccessible(true)
      method =>
        // We create a MethodHandles.Lookup that is allowed to look up private methods
        methodHandlesLookupConstructor
          .newInstance(method.getDeclaringClass, Integer.valueOf(MethodHandles.Lookup.PRIVATE))
          // Now using unreflect special, we get the default method from the declaring class, rather than the proxy
          .unreflectSpecial(method, method.getDeclaringClass)
} ...
...
```
but for other java versions it can create method handle for default method only if it returns `Descriptor` and has no parameters:
```
...
  } else {
      val lookup     = MethodHandles.lookup
      val returnType = MethodType.methodType(classOf[Descriptor])
      method => lookup.findSpecial(method.getDeclaringClass, method.getName, returnType, method.getDeclaringClass)
}
```
And if our default method not like this (`serviceName` in my example, or `withOpenAPI` in real case that @ihostage pointed was), it leads to `NoSuchMethodError`